functions/lib/crowd.js

```diff
@@ -0,0 +1,233 @@
+   * @param uid {string} Firebase UID, i.e. doc ID of appUsers collections
+   * @return {Promise<void>}
+   */
+  async function updateCrowdUser (uid) {
+    // Initialize some attributes to be empty, so to implicitly unset them on
+    // Crowd if the corresponding value is not found in the DB. Crowd expects
+    // an array for attribute values.
+    // FIXME: what happens if a user unlinks their crowd account first? We
+    //  should be resetting all attributes (e.g., remove the github_id), but for
+    //  this we need the removed crowd username. This can be found in the
+    //  Firestore document change snapshot (old value).
+    const attributeMap = {
+      github_id: []
+    }
+    return db.collection('appUsers')
+      .doc(uid)
+      .collection('accounts')
+      .get()
+      .then(query => {
+        const accounts = query.docs.map(d => d.data())
+        let crowdUsername = null
+        accounts.forEach(a => {
+          switch (a.hostname) {
+            case onfHostname:
+              crowdUsername = a.username
+              break
+            case 'github.com':
+              attributeMap.github_id = [a.username]
+              break
+            default:
+              console.log(`Unrecognized account hostname ${a.hostname} for uid ${uid}, ignoring`, a)
+              break
+          }
+        })
+        if (crowdUsername == null) {
+          console.log(`Cannot find Crowd account for uid ${uid}, aborting update`, accounts)
+          return
+        }
+        // Transform attributeMap in an object understood by Crowd.
+        const attributes = []
+        Object.keys(attributeMap).forEach(name => {
+          attributes.push({
+            name: name,
+            values: attributeMap[name]
+          })
+        })
+        console.log(`Pushing attribute for crowd user ${crowdUsername}`, attributes)
+        return setUserAttribute(crowdUsername, attributes)
+      })
+      .catch(console.error)
+  }
+
+  async function createSession (username, password) {
+    return rp.post({
+      ...rpConf,
+      uri: baseUri + '/session',
+      body: {
+        username: username,
+        password: password
+      }
+    })
+  }
+
+  async function invalidateSession (token) {
+    return rp.del({
+      ...rpConf,
+      uri: baseUri + `/session/${token}`
+    })
+  }
+
+  async function getUser (username) {
+    return rp.get({
+      ...rpConf,
+      uri: baseUri + `/user?username=${username}`
+    })
+  }
+
+  async function setUserAttribute (username, attributes) {
+    return rp.post({
+      ...rpConf,
+      uri: baseUri + `/user/attribute?username=${username}`,
+      body: {
+        attributes: attributes
+      }
+    })
+  }
+
+  async function getUsersWithGithubID (group) {
+    const validUsers = []
+    let users = {}
+    try {
+      users = await getUsersUnderGroup(group)
+    } catch (e) {
+      throw new functions.https.HttpsError('Listing user failed' + e)
```

Comment:
It doesn't make sense to throw `https.HttpsError` here. This is not a callable function, we should simply throw error.
```suggestion
      throw new Error('Listing user failed: ' + e)
```